fix: Switch to unittest.mock from mock

s-t-e-v-e-n-k opened the pull request:
Now that the minimum supported version of Python is 3.7, we can stop using the external mock requirement, and import it from unittest. I have also attempted to keep imports ordered.

Fixes #377
Closes #676

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-api-core/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

s-t-e-v-e-n-k committed:
test: Switch to unittest.mock from mock

Now that the minimum supported version of Python is 3.7, we can stop
using the external mock requirement, and import it from unittest. I have
also attempted to keep imports ordered.

Fixes #377

ohmayr commented:
@s-t-e-v-e-n-k, general comment: `AsyncMock` isn't part of the stdlib i.e. `unitest.mock` for Python 3.7. which we're using in our test code.

s-t-e-v-e-n-k commented:
Ahh, indeed, AsyncMock was added in 3.8. I'll modify this accordingly.

s-t-e-v-e-n-k committed:
test: Fallback to external mock for AsyncMock

AsyncMock is not included in unittest.mock under Python 3.7, so we must
fallback to the external mock requirement for that Python version. Only
install it for that version.

Keep this as a separate commit so it can be reverted when 3.7 isn't
supported anymore.

parthea committed:
Merge branch 'main' into switch-to-unittest-mock

parthea committed:
lint

parthea committed:
clean up to satisfy mypy

parthea committed:
lint

parthea committed:
fix build
